### PR TITLE
Method specific data structure

### PR DESCRIPTION
**Pull request by robinmarin**
Created an aem specific data structure, with create- and free-function.

### CODE REVIEW

File: src/solvers/aem/SimInf_solver_aem.c [hunk 1]
```diff
@@ -298,32 +322,164 @@ static int SimInf_solver_aem(
...
+ * @param out the resulting data structure.
+ * @param model structure with data about the model
+ * @param Nthread the number of threads available
+ * @param rng random number generator.
+ * @return 0 or SIMINF_ERR_ALLOC_MEMORY_BUFFER
+ */
+static int SimInf_aem_arguments_create(
+    SimInf_aem_arguments **out, SimInf_compartment_model *model, int Nthread, gsl_rng *rng)
+{
+    int i, error;
```
Comment (stewid) [inline src/solvers/aem/SimInf_solver_aem.c]: `error` can be removed since its value is never used

File: src/solvers/aem/SimInf_solver_aem.c [hunk 2]
```diff
@@ -298,32 +322,164 @@ static int SimInf_solver_aem(
...
+        goto on_error;
+
+    for (i = 0; i < Nthread; i++) {
+        int node;
+        SimInf_compartment_model *m = &model[i];
+        /* Binary heap storing all reaction events */
+        /* we have one for each node. Heap is thus only the size of the # transitions */
+        method[i].reactHeapSize = m->Nt;
+        method[i].reactNode = malloc(m->Nn * m->Nt * sizeof(int));
+        if (!method[i].reactNode) {
```
Comment (stewid) [inline src/solvers/aem/SimInf_solver_aem.c]: We can simplify this
```
if (!method[i].reactNode)
    goto on_error;
```

File: src/solvers/aem/SimInf_solver_aem.c [hunk 3]
```diff
@@ -298,32 +322,164 @@ static int SimInf_solver_aem(
...
+            goto on_error;
+        }
+
+        method[i].reactHeap = malloc(m->Nn * m->Nt * sizeof(int));
+        if (!method[i].reactHeap) {
+            error = SIMINF_ERR_ALLOC_MEMORY_BUFFER;
+            goto on_error;
+        }
+
+        method[i].reactTimes = malloc(m->Nn * m->Nt * sizeof(double));
```
Comment (stewid) [inline src/solvers/aem/SimInf_solver_aem.c]: See above

File: src/solvers/aem/SimInf_solver_aem.c [hunk 4]
```diff
@@ -298,32 +322,164 @@ static int SimInf_solver_aem(
...
+        /* we have one for each node. Heap is thus only the size of the # transitions */
+        method[i].reactHeapSize = m->Nt;
+        method[i].reactNode = malloc(m->Nn * m->Nt * sizeof(int));
+        if (!method[i].reactNode) {
+            error = SIMINF_ERR_ALLOC_MEMORY_BUFFER;
+            goto on_error;
+        }
+
+        method[i].reactHeap = malloc(m->Nn * m->Nt * sizeof(int));
+        if (!method[i].reactHeap) {
```
Comment (stewid) [inline src/solvers/aem/SimInf_solver_aem.c]: See above

File: src/solvers/aem/SimInf_solver_aem.c [hunk 5]
```diff
@@ -298,32 +322,164 @@ static int SimInf_solver_aem(
...
+        }
+
+        method[i].reactTimes = malloc(m->Nn * m->Nt * sizeof(double));
+        if (!method[i].reactTimes) {
+            error = SIMINF_ERR_ALLOC_MEMORY_BUFFER;
+            goto on_error;
+        }
+
+        method[i].reactInf = calloc(m->Nn * m->Nt, sizeof(double));
+        if (!method[i].reactInf) {
```
Comment (stewid) [inline src/solvers/aem/SimInf_solver_aem.c]: See above

File: src/solvers/aem/SimInf_solver_aem.c [hunk 6]
```diff
@@ -298,32 +322,164 @@ static int SimInf_solver_aem(
...
+            error = SIMINF_ERR_ALLOC_MEMORY_BUFFER;
+            goto on_error;
+        }
+
+        /* random generator for sample select with 1 per transition in each node */
+        method[i].rng_vec = malloc(m->Nn * m->Nt * sizeof(gsl_rng*));
+        if (!method[i].rng_vec) {
```
Comment (stewid) [inline src/solvers/aem/SimInf_solver_aem.c]: See above

File: src/solvers/aem/SimInf_solver_aem.c [hunk 7]
```diff
@@ -298,32 +322,164 @@ static int SimInf_solver_aem(
...
+            error = SIMINF_ERR_ALLOC_MEMORY_BUFFER;
+            goto on_error;
+        }
+
+        for (node = 0; node < m->Nn; node++) {
+            int trans;
+            for (trans = 0; trans < m->Nt; trans++) {
+                /* Random number generator */
+                method[i].rng_vec[m->Nt * node + trans] = gsl_rng_alloc(gsl_rng_mt19937);
+                if (!method[i].rng_vec[m->Nt * node + trans]) {
```
Comment (stewid) [inline src/solvers/aem/SimInf_solver_aem.c]: See above

File: src/solvers/aem/SimInf_solver_aem.c [hunk 8]
```diff
@@ -298,32 +322,164 @@ static int SimInf_solver_aem(
...
+                    error = SIMINF_ERR_ALLOC_MEMORY_BUFFER;
+                    goto on_error;
+                }
+                if (!method[i].rng_vec[m->Nt * node + trans]) {
```
Comment (stewid) [inline src/solvers/aem/SimInf_solver_aem.c]: See above

**Comment by robinmarin** [comment]
Addressed comments

**Comment by stewid** [comment]
Thanks

**Pull request merged**